### PR TITLE
Reduce redundant reflection

### PR DESCRIPTION
**Pull request by dmos62**
Related #2544

Should provide noticable speed up (and reduction in new Postgres connections being spawned) when logging in new demo users when there are a lot of databases.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: demo/middleware.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def __call__(self, request):
             engine = create_mathesar_engine(db_name)
             customize_settings(engine)
             load_custom_explorations(engine)
+            engine.dispose()
```
Comment (dmos62) [inline demo/middleware.py:42]: A relatively insignificant change.

File: demo/management/commands/clean_stale_db.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def drop_all_stale_databases(force=False, max_days=3, *args, **kwargs):
             if dropped:
                 deleted_databases.append(database.name)
                 database.delete()
-    reflect_db_objects(get_empty_metadata())
+                reflect_db_objects(get_empty_metadata(), db_name=database.name)
```
Comment (dmos62) [inline demo/management/commands/clean_stale_db.py:62]: Before we were reflecting Django models for every database, irrespective of what the clean up did. Now we'll be reflecting only the databases we affected.

File: mathesar/state/django.py [hunk 3]
```diff
@@ -53,9 +52,9 @@ def reflect_db_objects(metadata, db_name=None):
             models.Schema.current_objects.filter(database=database).delete()
 
 
-def sync_databases_status():
+def sync_databases_status(databases):
```
Comment (dmos62) [inline mathesar/state/django.py:55]: Before we were reflecting the status of every database, irrespective of whether or not we're intending to reflect only a single database. Now we'll only check the status of the database we've singled out.

File: mathesar/state/django.py [hunk 4]
```diff
@@ -32,11 +32,10 @@ def clear_dj_cache():
 
 
 def reflect_db_objects(metadata, db_name=None):
-    sync_databases_status()
     databases = models.Database.current_objects.all()
     if db_name is not None:
         databases = databases.filter(name=db_name)
-
+    sync_databases_status(databases)
```
Comment (dmos62) [inline mathesar/state/django.py:38]: Before we were reflecting the status of every database, irrespective of whether or not we're intending to reflect only a single database. Now we'll only check the status of the database we've singled out (when we've singled a database out).

**Review by mathemancer — APPROVED** [review]
I think this works, and should give us some performance benefits. Nice work.

**Pull request merged**